### DISCUSSION TITLE
One way to run UTM VMs in external drives.

### DISCUSSION THREAD

**Discussion by indicator0** [utmapp/UTM#3355] [Ideas]
UTM VMs now only run in internal system drives (~/Library/Containers/com.utmapp.UTM/Data/Documents). However, my storage is critical and cannot provide enough space to store .utm files. ( I believe if you are experiencing the same problem! Otherwise you won't  seeing this.) 
Here I find out a way to solve this problem in a not-that-smart way by using sparse image .A drawback is you must manually mount disk image before using UTM, but you may write a script to do this. 

0. Remember to close you UTM app.

1. Open **Disk Utility.app** and create a new blank sparse disk image and save it on your external drive. For SSD, APFS is recommended while HDD might experience less issues with HFS+ (Mac OS Extended).

2. Open **Terminal** and mount this image at the destination folder by running **hdiutil attach /Volumes/<externalDriveName>/<diskImageName>.sparseimage -mountpoint ~/Library/Containers/com.utmapp.UTM/Data/Documents** in terminal.

3. Now you may see a mounted image on your desktop or Finder sidebar. Move all files in 
~/Library/Containers/com.utmapp.UTM/Data/Documents to your new disk image.

4. Delete the original .utm file, launch your UTM app and your VM now runs in external drive.

Tested with UTM 2.4.1(40) on macOS Monterey 12.0.1 (21A559)

**Comment by usr-sse2** [comment]
Image is an unnecessary abstraction level which may slow down the performance. You can just mount a volume into the `~/Library/Containers/com.utmapp.UTM/Data/Documents` mount point using the `-mountPoint` option of `diskutil mount`.

**Reply by kingson87** [reply]
`hdiutil attach /Volumes/SSD/UTM.sparseimage -mountpoint ~/Library/Containers/com.utmapp.UTM/`

This command worked for me on M1 mac and copying the data folder as a whole

**Comment by DrBanksy** [comment]
When mounting the image I am getting this error:

hdiutil: attach failed - Resource busy

**Reply by yuvrajIMS** [reply]
First, you can simply build the VM on your local machine, after that restart the UTM. On the top right of the window, click the box with the up-pointing arrow, and from there you can select where to locate that VM.

**Comment by insung3511** [comment]
Sadly windows 11 doesn't work. Is it only happened to me? 

- macOS 13.3.1
- Macbook Pro M1

**Comment by 1sm23** [comment]
1. right click the VM
2. click move
3. select the external drive path

**Reply by jr3us** [reply]
Definitely the answer!

**Comment by anuprshetty** [comment]
Hey guys,
I got a simple straight forward solution to this.

I am giving the steps from the beginning:

1. install UTM on Mac
2. create required VMs in the UTM following the standard procedures.
--> at this point, created VMs complete data is stored at location --> ~/Library/Containers/com.utmapp.UTM/Data/Documents/
--> for example, in my case, i have created kali and parrot VMs. So data location of these VMs will be,
--> i) ~/Library/Containers/com.utmapp.UTM/Data/Documents/kali.utm (contains OS + data)
--> ii) ~/Library/Containers/com.utmapp.UTM/Data/Documents/parrot.utm (contains OS + data)
--> So ideally you just need to move above 2 utm files to your desired destination location. But don't do it manually. Let UTM software do the job. See next step.
3. Insert your external harddrive to Mac.
4. Now open your UTM dashboard(App). In the left side you see all the VMs.
5. Right click on each VM, then select "move" option. After that select the destination location as your external harddrive location. Then click "confirm".
6. Now you have moved your VM to your destination location. Also VM data at location ~/Library/Containers/com.utmapp.UTM/Data/Documents/ is removed. Repeat the same for all the VMs.


NOTE: UTM software simply needs an .utm file. So you can manually keep that file anywhere you want. Open UTM software dashboard. Then open finder and double click on the file which you have copied to open in UTM software. That's it.


Cons: You have to do this manually for each VM you create. Anyways we achieved what we want. Peace :)

**Comment by ralvarez83** [comment]
Hi.

I have found a simpler solution. You can create an Alias of the folder on your external disk, rename it to "Documents" and replace the "Documents" folder with this Alias.

I hope its helps.

**Comment by keekevin** [comment]
Hi, when I try to mount the Documents folder from terminal, I get "Permission denied".
 how can I fix it?

**Comment by malhal** [comment]
Even if you right click and move it to an external, if you then right click to clone then it copies it back to the internal, infuriating!

**Comment by ks09271** [comment]
Mounted /Users/ks/Library/Containers/com.utmapp.UTM form another volume doest'n work anymore under Tahoe Beta (macOS 26.0, 25A5346a)
`...
System Integrity Protection: disabled

Triggered by Thread: 0

Exception Type:    EXC_BREAKPOINT (SIGTRAP)
Exception Codes:   0x0000000000000001, 0x00000001a5daf0c8

Termination Reason:  Namespace SIGNAL, Code 5, Trace/BPT trap: 5
Terminating Process: exc handler [52860]


Application Specific Signatures:
failed integrity checks for container at /Users/ks/Library/Containers/com.utmapp.UTM/Data: Invalid argumen
...`
(Full log on request)

No idea yet, how to get rid of the problem (didn't happen with macOS 15 Sequoia, even the latest Beta)..

**Reply by osy** [reply]
> System Integrity Protection: disabled

SIP disabled is not a supported configuration by apple or by us

**Reply by ks09271** [reply]
Ah, thanks! But changing SIP has absolutely no effect on the problem!

**Reply by osy** [reply]
When you turn off SIP, you can make changes to the filesystem that break the system.

> failed integrity checks for container at /Users/ks/Library/Containers/com.utmapp.UTM/Data

This indicates it's a filesystem issue. The only way to "recover" is to re-install macOS

**Reply by ks09271** [reply]
The filesystem is OK (*) and used for other VMs, up and running (under qemu w/o UTM). Maybe the best "solution" for me is to scrap UTM and use qemu w/o the "decoration". 

(*) it's a ZFS-FS, not supported by anyone but works flowlessly since years - exept for UTM since "upgrading" to macOS 26.

**Comment by JC-000** [comment]
There is not a good reason for UTM to park the VM's files in a location like ~/Library/Containers/com.utmapp.UTM/Data/Documents. Normal macOS applications do not store files a user will manipulate in ~Library. It's a home for configurations, caches, etc and putting disk images in this location is a recipe for a user to not understand where their disk space is being consumed, and it makes it annoying if you want to move VM's between machines or the like. VMware Fusion and others follow the normal convention for file locations FWIW.

**Reply by osy** [reply]
There is indeed a good reason and that's because it's the only path that sandboxed apps like UTM are allowed to place data unrestricted. VMWare and others are not sandboxed.

**Reply by JC-000** [reply]
Hmm, I have not had an issue with moving the VM file to ~/Documents and running from there. I'll exercise this a bit.  I did see an issue where running from an external USB would work but as soon as I put a macOS VM in the background it would seemingly be immediately disconnected from storage and halt.

**Reply by osy** [reply]
Correct, when you move it/open from a different location, you give macOS permission for UTM to access that item. However, sandboxed apps do not have free reign to the filesystem. That's why for example, in theory VMWare and Parallels can steal all your data if they want to and UTM cannot (unless you explicitly add your files to a shared directory or something).

**Reply by JC-000** [reply]
Apple would not grant the requisite entitlement for the the App Store version either?

**Reply by osy** [reply]
There's these https://developer.apple.com/library/archive/documentation/Miscellaneous/Reference/EntitlementKeyReference/Chapters/AppSandboxTemporaryExceptionEntitlements.html but Apple requires you to have a good reason why you need it. I've never tried because App Sandbox is good security and there's no point in carving holes in it if you don't have to. For example, if you get malware on a VMWare VM and the malware is able to hack the hypervisor (rare but not unheard of), then that malware can access all the files on the host. With UTM, even if the malware can hack the hypervisor, it will still need to defeat macOS sandbox.

**Reply by osy** [reply]
If you want to place the VM somewhere else (like in Documents), there's nothing stopping you from doing so. Using the App Sandbox is being a good Mac citizen and is the way Apple has recommended apps work for a decade now (yet most don't follow the recommendations...)

**Reply by JC-000** [reply]
I appreciate the explanation and I do appreciate the security argument. Certainly virtualization has not proven to be magic and even if the hypervisor is somehow perfect the guest tools might not be etc. I would prefer to be able to carry some VMs on external media so having them hard power off and disconnect from virtual storage when they are backgrounded is a bit problematic. I have run into issues in the past with things storing large caches they never cleared in ~/Library structures so I've become wary of what I would think of as user files that are not ephemeral or modest in size being stuck in what is normally a hidden folder.
